feat(copy): adds on_copy to datasources to alter dynamic values

nlzet opened the pull request:
Add a on_copy feature for datasources to optionally alter the dynamic option answer value and label

anehx reviewed (approved):
Minor nitpick, otherwise looking good.

nlzet commented on caluma/caluma_data_source/data_sources.py:
```diff
@@ -86,3 +86,6 @@ def try_get_data_with_fallback(self, user, question, context):
         return new_data
+
+    def on_copy(self, old_answer, new_answer, old_value):
+        return old_value
```
Added typing, had to do a slight syntax change for python 3.9 support

```python
    def on_copy(
        self, old_answer: Answer, new_answer: Answer, old_value: Tuple[str, str]
    ) -> Tuple[Optional[str], Optional[str]]:
```

winged replied:
@nlzet If you want, you can revert the syntax hack for 3.9 - we decided that with Caluma 11, we don't support 3.9 anymore. You'll probably have to rebase onto the current `main` to drop the CI checks tho

nlzet replied:
@winged rebased and dropped the Optional syntax

winged commented on caluma/caluma_data_source/data_sources.py:
```diff
@@ -86,3 +87,8 @@ def try_get_data_with_fallback(self, user, question, context):
             return self.default
         return new_data
+
+    def on_copy(
+        self, old_answer: Answer, new_answer: Answer, old_value: Tuple[str, str]
+    ) -> Tuple[Optional[str], Optional[str]]:
+        return old_value
```
Can I have a docstring here to describe what's happening / what's supposed to happen in subclasses?

Or just update the class's docstring instead: What can this be useful for? What considerations to take into account? How to implement a certain behaviour?

winged commented on caluma/caluma_data_source/tests/data_sources.py:
```diff
@@ -91,3 +93,30 @@ def get_data(self, user, question, context):
+    def get_change_value(self):
+        return ["changedv-4b16-4a72-890b-cdbdc7da9f2", "changed label2"]
+
+    def on_copy(
+        self, old_answer: Answer, new_answer: Answer, old_value: Tuple[str, str]
+    ) -> Tuple[Optional[str], Optional[str]]:
+        old_slug, _ = old_value
```
I know it's "just" a test, but as those often are used as examples, I'd like to see a docstring here as well

winged commented on caluma/caluma_form/models.py:
```diff
@@ -553,6 +570,23 @@ def unlink_unused_rows(self, docs_to_keep):
             ans_doc.document.set_family(ans_doc.document)
             ans_doc.delete()
 
+    def modify_changed_choice_answer(self, question, answer_value, old_slug, new_slug):
```
Docstring please :-)

winged commented on caluma/caluma_form/tests/test_answer.py:
```diff
@@ -513,3 +513,60 @@ def test_file_answer_mutation_update_missing_file(
+        ),
+    ],
+)
+def test_modify_changed_choice_answer(
+    db,
+    question_factory,
+    answer_factory,
+    question_type,
+    answer_value,
+    old_slug,
+    new_slug,
+    expected,
+):
+    question = question_factory(type=question_type)
+    answer = answer_factory(question=question, value=answer_value)
+
+    assert (
```
I've started doing this for my own code, but I'd like to see more of it: Docstring on tests as well: Document the assumptions and expected overall behaviour

winged commented on caluma/caluma_form/tests/test_document.py:
```diff
@@ -1362,6 +1369,176 @@ def test_copy_document(
+    "on_copy_result",
+    ["discard", "change", None],
+)
+def test_copy_document_datasource_on_copy(
+    db,
+    document_factory,
+    answer_factory,
+    dynamic_option_factory,
+    form_factory,
+    form_question_factory,
+    schema_executor,
+    settings,
+    on_copy_result,
+):
```
Same here - document expected behaviour as a rough outline

winged reviewed (changes requested):
Nice work! I'd like to see more documentation however, and please also expand the commit message a bit to explain what the change is intended to do

winged commented on caluma/caluma_form/tests/test_document.py:
```diff
@@ -1362,6 +1369,196 @@ def test_copy_document(
+        The single dynamic question answer stays unchanged, the dynamic option is copied unchanged.
+        The multiple dynamic question answer stays unchanged. The dynamic options are copied unchanged.
+
+    """
```
Same agin - please reformat to use 88chars width maximum

nlzet commented:
@winged thanks for the updates. I added docstrings, reformatted for max 88chars width and dropped the python 3.9 syntax support. Could you have another look ?

winged reviewed (approved):
Nice work! :rocket:

nlzet commented:
@winged rebase fixed the tests due to https://github.com/projectcaluma/caluma/pull/2399
